cli: Add service subcommand

cathay4t opened the pull request:
With command `sudo nmstatectl service -c <config_folder>`, nmstatectl
will search all `.yml` files in specified folder(default to
`/etc/nmstate`), and apply those state files. Any applied state file
will be then renamed to `.applied` suffix.

The config files are sorted by rust native lexicographical order.

Man page for `nmstatectl` updated.
Add new man page `nmstate.service(8)`.
Systemd unit file included.
RPM spec file updated.
Integration test case included.

qinqon commented:
> With command `sudo nmstatectl service -c <config_folder>`, nmstatectl will search all `.yml` files in specified folder(default to `/etc/nmstate`), and apply those state files. Any applied state file will be then renamed to `.applied` suffix.
> 

I would reuse `apply` verb but if instead of a file is a directory it will apply or the files there on lexicographical order and add an option that will mark the files as applied if needed and no defaults since we have the systemd unit with proper options and will simplify stuff.

So the systemd unit will do 

`nmstatectl apply /etc/nmstate/state.d/ --rename-after-apply`

What do you think ?

dustymabe commented:
Hi. Is the `/etc/nmstate` location something people are already set on? Could I propose `/etc/NetworkManager/nmstate` instead?

cathay4t commented:
> > With command `sudo nmstatectl service -c <config_folder>`, nmstatectl will search all `.yml` files in specified folder(default to `/etc/nmstate`), and apply those state files. Any applied state file will be then renamed to `.applied` suffix.
> 
> I would reuse `apply` verb but if instead of a file is a directory it will apply or the files there on lexicographical order and add an option that will mark the files as applied if needed and no defaults since we have the systemd unit with proper options and will simplify stuff.
> 
> So the systemd unit will do
> 
> `nmstatectl apply /etc/nmstate/state.d/ --rename-after-apply`
> 
> What do you think ?

With ` --rename-after-apply`, I have to handle it been used along with `--memory-only`, `--no-commit` options, which clearly not used for this use case. If not do so, I have to explain why this `--rename-after-apply` cannot be used with other options.

So, I prefer `service` subcommand over `--rename-after-apply` argument.

cathay4t commented:
> Hi. Is the `/etc/nmstate` location something people are already set on? Could I propose `/etc/NetworkManager/nmstate` instead?

It is not decided yet. But `/etc/NetworkManager/nmstate` indicate we are sub-project of NetworkManager which this project initially was targeting to hide the backend details. And there are `ovsdb` features not supported by NetworkManger but supported by nmstate via `ovsdb` plugin.

qinqon commented on tests/integration/nmstate_service_test.py:
```diff
@@ -0,0 +1,79 @@
+"""
+
+CONFIG_DIR = "/etc/nmstate"
+TEST_CONFIG_FILE_NAME = "nmstate-test.yml"
```
Can we test with a pair of files like `'01-nmstate-test.yml` and `02-nmstate-test.yml`, and also to be sure that they are being applied at proper order 01 can create a "dummy" and 02 can modify that dummy so if they are applied at correct order it will work.

cathay4t replied:
Sure. Let me do that.

cathay4t committed:
cli: Add service subcommand

With command `sudo nmstatectl service -c <config_folder>`, nmstatectl
will search all `.yml` files in specified folder(default to
`/etc/nmstate`), and apply those state files. Any applied state file
will be then renamed to `.applied` suffix.

The config files are sorted by rust native lexicographical order.

Man page for `nmstatectl` updated.
Add new man page `nmstate.service(8)`.
Systemd unit file included.
RPM spec file updated.
Integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

jcaamano commented:
I wonder if we want to make the distinction of system or default state files and user state files

A system or default state file would go into `/etc/nmstate/<well-known-name>.yaml`
User state files would go into `/etc/nmstate/conf.d/*.yaml` and be processed in lexicographical order.

cathay4t commented:
> I wonder if we want to make the distinction of system or default state files and user state files
> 
> A system or default state file would go into `/etc/nmstate/<well-known-name>.yaml` User state files would go into `/etc/nmstate/conf.d/*.yaml` and be processed in lexicographical order.

Any special use case required for this system state file?

jcaamano commented:
In openshift, I see how the installer may generate a very simple default  state file (to put a default interface in a bridge). The is nothing preventing the user to deploy other state files, either to configure other interfaces or to override some or all of what that generated state file does. So way of organizing this might be good.

cathay4t commented:
You may use `/etc/nmstate/00-installer.conf` and user can use other name. Does that also solve your problem?
